File: test/r/large_test.R

# ---- deps ----
library(fastnntr)
library(ggplot2)
library(ggrepel)
library(dplyr)
library(tidyr)

plot_fastnnt_from_matrix <- function(dist_mat,
                                     labels = NULL,            # character vector or NULL
                                     out_png = "plots/fast_nnt_neighbornet.png",
                                     out_svg = "plots/fast_nnt_neighbornet.svg",
                                     label_leaves_only = TRUE,
                                     width_by_weight = TRUE) {

  # 1) run NeighborNet (your Rust uses column names when labels = NULL)
  nx <- run_neighbour_net(dist_mat, labels, "")

  # 2) pull data via $ methods (no wrappers)
  tr    <- as.data.frame(nx$get_node_translations())     # id (1-based), label
  pos   <- as.data.frame(nx$get_node_positions())        # id, x, y
  edges <- as.data.frame(nx$get_graph_edges())           # edge_id, u, v, split_id, weight

  # 3) keep only edges with coords for both endpoints
  edges <- edges |>
    semi_join(pos, by = c("u" = "id")) |>
    semi_join(pos, by = c("v" = "id"))

  # 4) leaves = degree 1
  deg <- bind_rows(transmute(edges, id = u),
                   transmute(edges, id = v)) |>
    count(id, name = "deg")

  nodes <- pos |>
    left_join(tr,  by = "id") |>
    left_join(deg, by = "id") |>
    mutate(deg = replace_na(deg, 0L))

  leaves <- dplyr::filter(nodes, deg == 1L)

  # 5) edge coordinates for plotting
  edges_xy <- edges |>
    left_join(rename(pos, xu = x, yu = y), by = c("u" = "id")) |>
    left_join(rename(pos, xv = x, yv = y), by = c("v" = "id"))

  # 6) edge layer (optional width scaling by weight)
  edge_layer <-
    if (width_by_weight && nrow(edges_xy)) {
      rng <- range(edges_xy$weight, na.rm = TRUE)
      if (is.finite(rng[1]) && is.finite(rng[2]) && diff(rng) > 0) {
        edges_xy$width <- 0.3 + 1.7 * (edges_xy$weight - rng[1]) / diff(rng)
      } else {
        edges_xy$width <- 0.6
      }
      geom_segment(
        data = edges_xy,
        aes(x = xu, y = yu, xend = xv, yend = yv, linewidth = width),
        lineend = "round", alpha = 0.9, colour = "black", show.legend = FALSE
      )
    } else {
      geom_segment(
        data = edges_xy,
        aes(x = xu, y = yu, xend = xv, yend = yv),
        lineend = "round", linewidth = 0.6, alpha = 0.9, colour = "black"
      )
    }

  # 7) assemble plot (only leaf labels)
  p <- ggplot() +
    edge_layer +
    {
      if (label_leaves_only && nrow(leaves)) {
        list(
          geom_point(data = leaves, aes(x = x, y = y), size = 0.8, colour = "black"),
          ggrepel::geom_text_repel(
            data = leaves, aes(x = x, y = y, label = label),
            size = 2.2, box.padding = 0.05, point.padding = 0.05,
            segment.size = 0.1, max.overlaps = Inf
          )
        )
      } else NULL
    } +
    coord_equal() +
    theme_void() +
    ggtitle("Fast-NNT NeighborNet")

  dir.create(dirname(out_png), recursive = TRUE, showWarnings = FALSE)
  ggsave(out_png, p, width = 8, height = 8, dpi = 300)
  ggsave(out_svg, p, width = 8, height = 8)

  invisible(p)
}

# ---- example usage ----
# Build a toy symmetric distance matrix (labels taken from colnames)
set.seed(1)
# read in test/data/large/large_dist_matrix.csv
m <- as.matrix(read.csv("test/data/large/large_dist_matrix.csv", row.names = 1))
print(m)

plot_fastnnt_from_matrix(m)  # saves PNG+SVG into ./plots/

File: test/python/large_test.py
#! /usr/bin/env python

import fastnntpy as fn
import pandas as pd
import os
from collections import Counter
import networkx as nx
import numpy as np
import matplotlib.pyplot as plt
from matplotlib.collections import LineCollection


data = pd.read_csv("test/data/large/large_dist_matrix.csv")
n = fn.run_neighbor_net(data, max_iterations=5000, ordering_method="splitstree4", labels=data.columns)
n = fn.run_neighbour_net(data, max_iterations=5000, ordering_method="splitstree4", labels=data.columns)
print("Labels")
print(len(n.get_labels()))
print("Splits Records")
print(len(n.get_splits_records()))
print("Node Translations")
print(len(n.get_node_translations()))
print("Node Positions")
print(len(n.get_node_positions()))
print("Graph Edges")
print(len(n.get_graph_edges()))



def plot_fast_nnt_networkx(nx_obj, out_path="test/plots/fast_nnt_nx.png",
                           shift=0, node_size=10, font_size=7,
                           scale_width_by_weight=False, dpi=300):
    # -- data from PyNexus --
    labels = {i + shift: s for i, s in nx_obj.get_node_translations()}
    pos    = {i + shift: (x, y) for i, x, y in nx_obj.get_node_positions()}
    # corrected parsing order: (edge_id, u, v, sid, w)
    edges_raw = [ (u + shift, v + shift, w)
                  for (_eid, u, v, _sid, w) in nx_obj.get_graph_edges() ]

    # only keep edges whose endpoints have positions
    edges = [(u, v, w) for (u, v, w) in edges_raw if u in pos and v in pos]
    if not edges:
        raise ValueError("No drawable edges (endpoints missing positions).")

    # -- build graph --
    G = nx.Graph()
    for u, v, w in edges:
        G.add_edge(u, v, weight=w)

    # leaves only (degree == 1)
    leaves = [n for n, d in G.degree() if d == 1]

    # edge widths (optional)
    if scale_width_by_weight:
        ws = [G[u][v].get("weight", 1.0) for u, v in G.edges()]
        wmax = max(ws) if ws else 1.0
        widths = [0.5 + 2.5 * (w / wmax) for w in ws]
    else:
        widths = 0.8

    # -- draw (no layout) --
    os.makedirs(os.path.dirname(out_path) or ".", exist_ok=True)
    plt.figure(figsize=(8, 8), dpi=dpi)

    nx.draw_networkx_edges(G, pos, width=widths, edge_color="black", alpha=0.9)
    nx.draw_networkx_nodes(G, pos, nodelist=leaves, node_size=node_size, node_color="black")

    leaf_labels = {n: labels.get(n, str(n)) for n in leaves}
    nx.draw_networkx_labels(G, pos, labels=leaf_labels, font_size=font_size)

    plt.axis("equal"); plt.axis("off"); plt.tight_layout(pad=0.02)
    plt.savefig(out_path, dpi=dpi, bbox_inches="tight", pad_inches=0.01)
    base, _ = os.path.splitext(out_path)
    plt.savefig(base + ".svg", bbox_inches="tight", pad_inches=0.01)
    plt.close()
    return out_path

plot_fast_nnt_networkx(n, out_path="test/plots/fast_nnt_graph_networkx.png")

def save_neighbornet_plot(nx_obj,
                          out_path="test/plots/fast_nnt_neighbornet.png",
                          shift=0,
                          label_leaves=True,
                          scale_width_by_weight=True,
                          node_size=8,
                          font_size=6,
                          dpi=300):
    """
    Render NeighborNet as straight segments from Rust-provided positions.
    - Labels ONLY leaf nodes (degree==1).
    - No internal node markers, no edge labels.
    - Set `shift` if your IDs are 1/2-based etc.
    """
    # --- data from PyNexus / Rust ---
    labels = {i + shift: s for i, s in nx_obj.get_node_translations()}
    pos     = {i + shift: (x, y) for i, x, y in nx_obj.get_node_positions()}
    edges   = [(u + shift, v + shift, w) for (_eid, u, v, _sid, w) in nx_obj.get_graph_edges()]

    # keep only edges whose endpoints have positions
    edges = [(u, v, w) for (u, v, w) in edges if u in pos and v in pos]
    if not edges:
        raise ValueError("No drawable edges (endpoints missing positions).")

    # degrees → leaf detection
    deg = Counter()
    for u, v, _ in edges:
        deg[u] += 1; deg[v] += 1
    leaves = [n for n, d in deg.items() if d == 1]

    # segments for fast draw
    segs = [[pos[u], pos[v]] for (u, v, _w) in edges]

    # optional edge widths by weight
    if scale_width_by_weight:
        ws = np.array([w for *_ , w in edges], dtype=float)
        wmin, wptp = np.nanmin(ws), np.ptp(ws)
        widths = 0.5 + 2.5 * ((ws - wmin) / (wptp if wptp else 1.0))
    else:
        widths = 0.8

    # --- draw ---
    os.makedirs(os.path.dirname(out_path) or ".", exist_ok=True)
    fig, ax = plt.subplots(figsize=(8, 8), dpi=dpi)

    ax.add_collection(LineCollection(segs, colors="black", linewidths=widths, alpha=0.9))

    # draw & label leaves only
    if leaves:
        xs = [pos[n][0] for n in leaves]
        ys = [pos[n][1] for n in leaves]
        ax.scatter(xs, ys, s=node_size, c="black", zorder=3)
        for n in leaves:
            x, y = pos[n]
            ax.text(x, y, labels.get(n, str(n)),
                    fontsize=font_size, ha="center", va="center", zorder=4)

    ax.set_aspect("equal", adjustable="box")
    ax.set_xlabel("x"); ax.set_ylabel("y"); ax.set_title("Fast-NNT NeighborNet")
    ax.autoscale(); ax.margins(0.02); ax.grid(True, alpha=0.15)

    fig.savefig(out_path, bbox_inches="tight", pad_inches=0.02)
    fig.savefig(os.path.splitext(out_path)[0] + ".svg", bbox_inches="tight", pad_inches=0.02)
    plt.close(fig)
    return out_path

save_neighbornet_plot(n, "test/plots/fast_nnt_graph_matplotlib.png", shift=2)